python/src/spark_rapids_ml/dbscan.py

```diff
@@ -0,0 +1,747 @@
+        ),
+        typeConverter=TypeConverters.toString,
+    )
+
+    max_mbytes_per_batch = Param(
+        Params._dummy(),
+        "max_mbytes_per_batch",
+        (
+            f"Calculate batch size using no more than this number of megabytes for the pairwise distance computation."
+            f"This enables the trade-off between runtime and memory usage for making the N^2 pairwise distance computations more tractable for large numbers of samples."
+            f"If you are experiencing out of memory errors when running DBSCAN, you can set this value based on the memory size of your device."
+        ),
+        typeConverter=TypeConverters.toInt,
+    )
+
+    calc_core_sample_indices = Param(
+        Params._dummy(),
+        "calc_core_sample_indices",
+        (
+            f"Indicates whether the indices of the core samples should be calculated."
+            f"Setting this to False will avoid unnecessary kernel launches"
+        ),
+        typeConverter=TypeConverters.toBoolean,
+    )
+
+    idCol = Param(
+        Params._dummy(),
+        "idCol",
+        "id column name.",
+        typeConverter=TypeConverters.toString,
+    )
+
+    def getFeaturesCol(self) -> Union[str, List[str]]:  # type: ignore
+        """
+        Gets the value of :py:attr:`featuresCol` or :py:attr:`featuresCols`
+        """
+        if self.isDefined(self.featuresCols):
+            return self.getFeaturesCols()
+        elif self.isDefined(self.featuresCol):
+            return self.getOrDefault("featuresCol")
+        else:
+            raise RuntimeError("featuresCol is not set")
+
+    def setFeaturesCol(self: P, value: Union[str, List[str]]) -> P:
+        """
+        Sets the value of :py:attr:`featuresCol` or :py:attr:`featuresCols`.
+        """
+        if isinstance(value, str):
+            self._set_params(featuresCol=value)
+        else:
+            self._set_params(featuresCols=value)
+        return self
+
+    def setFeaturesCols(self: P, value: List[str]) -> P:
+        """
+        Sets the value of :py:attr:`featuresCols`. Used when input vectors are stored as multiple feature columns.
+        """
+        return self._set_params(featuresCols=value)
+
+    def setPredictionCol(self: P, value: str) -> P:
+        """
+        Sets the value of :py:attr:`predictionCol`.
+        """
+        self._set_params(predictionCol=value)
+        return self
+
+    def setIdCol(self: P, value: str) -> P:
+        """
+        Sets the value of `idCol`. If not set, an id column will be added with column name `unique_id`. The id column is used to specify nearest neighbor vectors by associated id value.
+        """
+        self._set_params(idCol=value)
+        return self
+
+    def getIdCol(self) -> str:
+        """
+        Gets the value of `idCol`.
+        """
+        return self.getOrDefault("idCol")
+
+    def _ensureIdCol(self, df: DataFrame) -> DataFrame:
```

Comment:
Can we move idCol, setter, getter, and _ensureIdCol to a new class and reuse the class in both NearestNeighbors and DBSCAN?